daemon/src/identify/dialer.rs

```diff
@@ -0,0 +1,176 @@
+use crate::identify::protocol;
+use crate::identify::PROTOCOL;
+use crate::Environment;
+use async_trait::async_trait;
+use std::collections::HashMap;
+use tokio_extras::spawn_fallible;
+use xtra::Address;
+use xtra::Context;
+use xtra_libp2p::endpoint;
+use xtra_libp2p::libp2p::PeerId;
+use xtra_libp2p::Endpoint;
+use xtra_libp2p::OpenSubstream;
+use xtra_productivity::xtra_productivity;
+
+pub struct Actor {
+    endpoint: Address<Endpoint>,
+    identities: HashMap<PeerId, IdentityInfo>,
+}
+
+impl Actor {
+    pub fn new(endpoint: Address<Endpoint>) -> Self {
+        NUM_LIBP2P_CONNECTIONS_GAUGE.reset();
+
+        Self {
+            endpoint,
+            identities: HashMap::default(),
+        }
+    }
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = ();
+
+    async fn stopped(self) -> Self::Stop {}
+}
+
+pub(crate) struct GetIdentifyInfo(pub PeerId);
+
+pub(crate) struct PeerIdentityReceived {
+    peer_id: PeerId,
+    peer_identity: protocol::IdentifyMsg,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct IdentityInfo {
+    pub wire_version: String,
+    pub daemon_version: String,
+    pub environment: Environment,
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("Conversion to identity info failed: {error}")]
+pub struct ConversionError {
+    #[source]
+    error: anyhow::Error,
+}
+
+impl TryFrom<protocol::IdentifyMsg> for IdentityInfo {
+    type Error = ConversionError;
+
+    fn try_from(identity_msg: protocol::IdentifyMsg) -> Result<Self, Self::Error> {
+        let identity_info = IdentityInfo {
+            wire_version: identity_msg.wire_version(),
+            daemon_version: identity_msg
+                .daemon_version()
+                .map_err(|error| ConversionError { error })?,
```

Comment:
No need to explicitly map if you annotate line 55 with `from` instead of `source`.

```suggestion
               ?,
```